### PR TITLE
feat(clickhousegrant): add new kind

### PR DESCRIPTION
**Pull request by rriski**
Add new `ClickhouseGrant` kind to control user and role grants.

### CODE REVIEW

**Comment by rriski** [comment]
[tests](https://github.com/aiven/aiven-operator/actions/runs/9518388781)

**Review by byashimov — CHANGES_REQUESTED** [review]

File: controllers/clickhousegrant_controller.go [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+		}
+	}
+
+	meta.SetStatusCondition(&g.Status.Conditions,
+		getInitializedCondition("Created",
+			"Instance was created or update on Aiven side"))
+
+	meta.SetStatusCondition(&g.Status.Conditions,
+		getRunningCondition(metav1.ConditionUnknown, "Created",
+			"Instance was created or update on Aiven side, status remains unknown"))
```
[the remark this answers is not in the document]
Comment (rriski) [reply controllers/clickhousegrant_controller.go:98]: We have the same status condition in many resources (e.g. kafkaacl, connectionpool etc). Should all of these be updated?

image

File: controllers/clickhousegrant_controller.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+		For(&v1alpha1.ClickhouseGrant{}).
+		Complete(r)
+}
+
+var (
+	// 'aiven', 'aiven_monitoring', and 'avnadmin' are Aiven-managed users.
+	internalAivenRoles = []string{"aiven", "aiven_monitoring", "avnadmin"}
+	// "OR user_name IS NULL" is needed to include roles in the result
+	allNonAivenUsersQuery = fmt.Sprintf("SELECT name FROM system.users WHERE name NOT IN('%s')", strings.Join(internalAivenRoles, "', '"))
+	allRolesQuery         = "SELECT name FROM system.roles"
```
[the remark this answers is not in the document]
Comment (rriski) [reply controllers/clickhousegrant_controller.go:54]: Using const results in error.

image

I feel like this isn't as readable as having them all as `var`'s:

```
var (
	// 'aiven', 'aiven_monitoring', and 'avnadmin' are Aiven-managed users.
	internalAivenRoles = []string{"aiven", "aiven_monitoring", "avnadmin"}
	// "OR user_name IS NULL" is needed to include roles in the result
	allNonAivenUsersQuery = fmt.Sprintf("SELECT name FROM system.users WHERE name NOT IN('%s')", strings.Join(internalAivenRoles, "', '"))
)
const (
	allRolesQuery         = "SELECT name FROM system.roles"
)
```

File: controllers/clickhousegrant_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (byashimov) [reply controllers/clickhousegrant_controller.go:98]: I mean, mysql creates the secret in `get()` that's why it says `status remains unknown`, but this resource knows its status.

[the remark this answers is not in the document]
Comment (rriski) [reply controllers/clickhousegrant_controller.go:98]: Ah now I understood. So you are suggesting to run a query to check if all the privileges have been created on Aiven's side in get. I agree with that approach, I'll add it

**Review by byashimov — DISMISSED** [review]
lgtm

**Review by byashimov — APPROVED** [review]

**Pull request merged**